packages/node_modules/@node-red/editor-client/src/js/ui/common/typedInput.js

```diff
@@ -108,7 +108,87 @@
             return matches;
         }
     }
-    
+
+    function getEnvVars (obj, envVars = {}) {
+        let parent
+        if (obj.type === 'tab' || obj.type === 'subflow') {
+            RED.nodes.eachConfig(function (conf) {
+                if (conf.type === "global-config") {
+                    parent = conf;
+                }
+            })
+        } else if (obj.g) {
+            parent = RED.nodes.group(obj.g)
+        } else if (obj.z) {
+            parent = RED.nodes.workspace(obj.z) || RED.nodes.subflow(obj.z)
+        }
+        if (parent) {
+            getEnvVars(parent, envVars)
+        }
+        if (obj.env) {
+            obj.env.forEach(env => {
+                envVars[env.name] = obj
+            })
+        }
+        return envVars
+    }
+    RED.pp = getEnvVars
+    const envAutoComplete = function (val) {
+        const editStack = RED.editor.getEditStack()
+        if (editStack.length === 0) {
+            done([])
+            return
+        }
+        const editingNode = editStack.pop()
+        if (!editingNode) {
+            return []
+        }
+        const envVarsMap = getEnvVars(editingNode)
+        const envVars = Object.keys(envVarsMap)
```

Comment:
If the environment variable type is JSON, we should inspect that as well

Reply:
We don't have a scenario where you need to evaluate inside the value of the env var - it isn't the same usage as context where you do want to reference inside the objects.